dali/kernels/imgproc/jpeg/jpeg_distortion_gpu.cuh

```diff
@@ -0,0 +1,255 @@
+}
+
+template <bool horz_subsample, bool vert_subsample, typename T>
+__inline__ __device__
+void ycbcr_to_rgb_subsampled(ivec2 offset, const Surface2D<uint8_t>& out,
+                             YCbCrSubsampled<T, horz_subsample, vert_subsample> ycbcr) {
+  int y = offset.y;
+  int x = offset.x;
+  write_vec(&out(x, y), ycbcr_to_rgb(vec<3, T>(ycbcr.luma[0], ycbcr.cb, ycbcr.cr)));
+  if (horz_subsample && vert_subsample) {
+    write_vec(&out(x + 1, y), ycbcr_to_rgb(vec<3, T>(ycbcr.luma[1], ycbcr.cb, ycbcr.cr)));
+    write_vec(&out(x, y + 1), ycbcr_to_rgb(vec<3, T>(ycbcr.luma[2], ycbcr.cb, ycbcr.cr)));
+    write_vec(&out(x + 1, y + 1), ycbcr_to_rgb(vec<3, T>(ycbcr.luma[3], ycbcr.cb, ycbcr.cr)));
+  } else if (horz_subsample) {
+    write_vec(&out(x + 1, y), ycbcr_to_rgb(vec<3, T>(ycbcr.luma[1], ycbcr.cb, ycbcr.cr)));
+  } else if (vert_subsample) {
+    write_vec(&out(x, y + 1), ycbcr_to_rgb(vec<3, T>(ycbcr.luma[1], ycbcr.cb, ycbcr.cr)));
+  }
+}
+
+template <bool horz_subsample, bool vert_subsample>
+__global__ void ChromaSubsampleDistortion(const SampleDesc *samples,
+                                          const kernels::BlockDesc<2> *blocks) {
+  using T = uint8_t;
+  const auto &block = blocks[blockIdx.x];
+  const auto &sample = samples[block.sample_idx];
+
+  int y_start = threadIdx.y + block.start.y;
+  int x_start = threadIdx.x + block.start.x;
+  if (y_start >= block.end.y || x_start >= block.end.x) {
+    return;
+  }
+
+  const Surface2D<const uint8_t> in = {
+    sample.in, sample.size, 3, sample.strides, 1
+  };
+
+  const Surface2D<uint8_t> out = {
+    sample.out, sample.size, 3, sample.strides, 1
+  };
+
+  for (int pos_y = y_start; pos_y < block.end.y; pos_y += blockDim.y) {
+    for (int pos_x = x_start; pos_x < block.end.x; pos_x += blockDim.x) {
+      int y = pos_y << vert_subsample;
+      int x = pos_x << horz_subsample;
+      auto ycbcr = rgb_to_ycbcr_subsampled<horz_subsample, vert_subsample, T>(ivec2{x, y}, in);
+      ycbcr_to_rgb_subsampled<horz_subsample, vert_subsample, T>(ivec2{x, y}, out, ycbcr);
+    }
+  }
+}
+
+
+
+template <bool horz_subsample, bool vert_subsample>
+__global__ void JpegCompressionDistortion(const SampleDesc *samples,
+                                          const kernels::BlockDesc<2> *blocks) {
+  using T = uint8_t;
+  const auto &block = blocks[blockIdx.x];
+  const auto &sample = samples[block.sample_idx];
+
+  int y_start = threadIdx.y + block.start.y;
+  int x_start = threadIdx.x + block.start.x;
+  if (y_start >= block.end.y || x_start >= block.end.x) {
+    return;
+  }
+
+  // Assuming CUDA block is 32x16, leading to a 2x4 grid of chroma blocks
+  // and up to 4x8 blocks of luma.
+  __shared__ T luma_blk[2 << vert_subsample][4 << horz_subsample][8][9];
+  __shared__ T cb_blk[2][4][8][9];  // 8+1 to reduce bank conflicts
+  __shared__ T cr_blk[2][4][8][9];  // 8+1 to reduce bank conflicts
+
+  int chroma_x = threadIdx.x & 7;  // % 8
+  int chroma_y = threadIdx.y & 7;  // % 8
```

Comment:
I checked and for both `& 7` and `% 8` NVCC generates the same code:
```
mov.u32         %r3, %tid.x;
and.b32         %r4, %r3, 7;
```
Maybe we should stick to what is intended here then? Unless there is something here I'm not getting.

This goes for every time this trick was used.

Reply:
@mzient This was originally your suggestion. What do you think?

Reply:
The same for `/8` vs `>>8`. For both it is:
```
mov.u32        %r3, %tid.x;
shr.u32         %r4, %r3, 3;
```

Reply:
My original suggestion was inside a function where these values were passed as signed integers. For signed integers, the division rounds towards zero and gives negative remainder. It's not only slower (additional math), but also potentially dangerous.